x/wasm/IBC.md

```diff
@@ -178,96 +178,98 @@ type IBCCloseChannel struct {
 type Version string
 ```
 
-Packet callbacks:
+Channel andPacket callbacks to the wasm contract
 
 ```go
-package packets
-
-// for reference: this is more like what we pass to go-cosmwasm
-// func (c *mockContract) OnReceive(params cosmwasm2.Env, msg []byte, store prefix.Store, api cosmwasm.GoAPI, 
-//         querier keeper.QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm2.OnReceiveIBCResponse, uint64, error) {}
-// below is how we want to expose it in x/wasm:
-
-// IBCRecvPacket is called when we receive a packet sent from the other end of a channel
-// The response bytes listed here will be returned to the caller as "result" in IBCPacketAck
-// What do we do with error?
-//
-// If we were to assume/enforce an envelope, then we could wrap response/error into the acknowledge packet,
-// but we delegated that encoding to the contract
-func IBCRecvPacket(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, env IBCPacketInfo, msg []byte) 
-    (response []byte, err error) {}
-
-// how to handle error here? if we push it up the ibc stack and fail the transaction (normal handling),
-// the packet may be posted again and again. just log and ignore failures here? what does a failure even mean?
-// only realistic one I can imagine is OutOfGas (panic) to retry with higher gas limit.
-//
-// if there any point in returning a response, what does it mean?
-func IBCPacketAck(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, env IBCPacketInfo, 
-    originalMsg []byte, result []byte) error {}
-
-// same question as for IBCPacketAck
-func IBCPacketDropped(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, env IBCPacketInfo, 
-    originalMsg []byte, errorMsg string) error {}
-
-// do we need/want all this info?
-type IBCPacketInfo struct {
-	// local id for the Channel packet was sent/received on
-	ChannelID string
 
-    // sequence for the packet (will already be enforced in order if ORDERED channel)
-	Sequence uint64
-	
-    // Note: Timeout if guaranteed ONLY to be exceeded iff we are processing IBCPacketDropped
-    // otherwise, this is just interesting metadata
+// IBCContractCallbacks defines the methods for go-cosmwasm to interact with the wasm contract.
+// A mock contract would implement the interface to fully simulate a wasm contract's behaviour.
+type IBCContractCallbacks interface {
+  // Package livecycle
+
+  // OnIBCPacketReceive handles an incoming IBC package
+  OnIBCPacketReceive(hash []byte, params cosmwasm.Env, msg []byte, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCPacketReceiveResponse, uint64, error)
```

Comment:
See comments at the end, please use this instead of `msg`

```go
type IBCPacket struct {
  Msg []byte
  Sequence uint64
  TimeoutHeight uint64
  // block timestamp (in nanoseconds) after which the packet times out
  TimeoutTimestamp uint64
}
```